File: LeetCodeRecord/MethodFirst/offer53-Search.java
package MethodFirst;

class Search {
    public int search(int[] nums, int target) {
        int i = 0;
        int j = nums.length - 1;

        // 先找右边界
        while (i <= j) {
            int mid = i + ((j - i) >> 1);
            if (nums[mid] <= target) {
                i = mid + 1;
            } else {
                j = mid - 1;
            }
        }
        int right = i;
        if(j < 0 || nums[j] != target) {
            return 0;
        }

        i = 0;
        j = nums.length - 1;
        // 找左边界
        while (i <= j) {
            int mid = i + ((j - i) >> 1);
            if (nums[mid] >= target) {
                j = mid - 1;
            } else {
                i = mid + 1;
            }
        }

        int left = j;

        return right - left - 1;
    }    
}

File: LeetCodeRecord/ACFirst-Golang/offer53-Search.go
package main

func search(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	low := 1
	high := 0

	for left <= right {
		mid := left + ((right - left) >> 1)

		if nums[mid] < target {
			left = mid + 1
		} else if nums[mid] > target {
			right = mid - 1
		} else {
			if mid == 0 || nums[mid-1] != target {
				low = mid
				break
			}
			right = mid - 1
		}
	}

	left = 0
	right = len(nums) - 1

	for left <= right {
		mid := left + ((right - left) >> 1)

		if nums[mid] < target {
			left = mid + 1
		} else if nums[mid] > target {
			right = mid - 1
		} else {
			if mid == len(nums)-1 || nums[mid+1] != target {
				high = mid
				break
			}

			left = mid + 1
		}
	}

	return high - low + 1
}
